diffrax/_integrate.py

```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
     dense_save_index: Optional[IntScalarLike]
 
 
-def _is_none(x):
+class _ShapeDtypeTypeCoercer(AbstractArray):
+    def __new__(cls, term_type: Any):
+        if not issubclass(term_type, AbstractArray):
+            return term_type
+
+        cls.array_type = jax.ShapeDtypeStruct
+        cls.dtypes = term_type.dtypes
+        cls.dims = term_type.dims
+        cls.index_variadic = term_type.index_variadic
+        cls.dim_str = term_type.dim_str
+        return cls
```

Comment:
Haha, I see what you're doing here! FWIW the attributes on `AbstractArray` aren't public API (maybe we should add some underscores in front) so this is a bit dangerous.

To check the shape and dtype of a struct against a jaxtyping array annotation e.g. `Float[Array, "foo"]`, then I'd probably use `jax.eval_shape` to cast the struct to an array:
```python
def _assert(x):
    assert isinstance(x, Float[Array, "foo"])
jax.eval_shape(_assert, struct)
```

Reply:
Really like your solution, much better than my hack haha!